### DISCUSSION TITLE
Support for gunicorn's callbacks worker_int and worker_abort

### DISCUSSION THREAD

**Discussion by marcinbarczynski** [Kludex/uvicorn#1688] [Ideas]
Uvicorn worker can exit in two ways:
- kindly (one SIGINT): wait for all tasks to complete and run lifespan's shutdown events
- forcibly (two SIGINTs): exit immediately

Gunicorn supports `worker_int` and `worker_abort` callbacks that are run when a worker receives SIGINT and SIGABRT respectively. Unfortunately, the callbacks are not supported by uvicorn. That makes it impossible to reliably run a piece of code when a worker exits.
How about adding supporting the gunicorn's callbacks?

**Comment by Kludex** [comment]
Well... The premise is not even valid when using `gunicorn`: https://github.com/benoitc/gunicorn/issues/2604.

Can you link how `gunicorn` supports `worker_int` and `worker_abort`?

**Reply by marcinbarczynski** [reply]
Here are the links to documentation:
- https://docs.gunicorn.org/en/latest/settings.html#worker-int
- https://docs.gunicorn.org/en/latest/settings.html#worker-abort

Honestly, it doesn't matter which signal is sent internally between master process and workers as long as it's possible to run the callback reliably.

**Reply by marcinbarczynski** [reply]
Any chance for putting this in the pipeline?
